compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -326,8 +326,23 @@ object Erasure {
     }
 
     private def checkNotErased(tree: Tree)(implicit ctx: Context): tree.type = {
-      if (isErased(tree) && !ctx.mode.is(Mode.Type))
-        ctx.error(em"${tree.symbol} is declared as erased, but is in fact used", tree.sourcePos)
+      if (!ctx.mode.is(Mode.Type)) {
+        if (isErased(tree))
+          ctx.error(em"${tree.symbol} is declared as erased, but is in fact used", tree.sourcePos)
+        tree.symbol.getAnnotation(defn.CompileTimeOnlyAnnot) match {
+          case Some(annot) =>
+            if (ctx.scala2Mode) {
+              // Scala 2 macros cannot be expanded for which we emit warnings. The best we can do is warn as well.
+              ctx.warning("Reference to member with @compileTimeOnly", tree.sourcePos)
+            } else {
+              annot.tree match {
+                case Apply(_, Literal(msg) :: Nil) => ctx.error(msg.stringValue, tree.sourcePos)
+                case _ => ctx.error("Reference to member with @compileTimeOnly", tree.sourcePos)
+              }
+            }
+          case _ => // OK
+        }
+      }
```

Comment:
In `RefChecks.scala`, there is commented code:

```Scala
    /*  (Not enabled yet)
       *  See an explanation of compileTimeOnly in its scaladoc at scala.annotation.compileTimeOnly.
       *
      if (sym.isCompileTimeOnly) {
        def defaultMsg =
          sm"""Reference to ${sym.fullLocationString} should not have survived past type checking,
              |it should have been processed and eliminated during expansion of an enclosing macro."""
        // The getOrElse part should never happen, it's just here as a backstop.
        ctx.error(sym.compileTimeOnlyMessage getOrElse defaultMsg, pos)
      }*/
```

Maybe we should just revive the code?

Reply:
No, it would not take into account if it is into an erased context or not. Which is already performed in the place where I check it.

Reply:
As in https://github.com/lampepfl/dotty/pull/6424/files#diff-a12cb93e6ac05566103dd86ede6e5901R12, where the reference of the `@compileTimeOnly` is inside an `erased def`